core/src/main/java/com/linecorp/armeria/common/logging/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,298 @@
+                                 .decActiveRequests(values);
+
+                    if (isSuccess(log)) {
+                        metricAdapter.success(values);
+                    } else {
+                        metricAdapter.failure(values);
+                    }
+                }, RequestLogAvailability.COMPLETE);
+        return function.apply(ctx, req);
+    }
+
+    @Override
+    public O serve(Service<I, O> delegate, ServiceRequestContext ctx, I req) throws Exception {
+        return request(delegate::serve, ctx, req);
+    }
+
+    @Override
+    public O execute(Client<I, O> delegate, ClientRequestContext ctx, I req) throws Exception {
+        return request(delegate::execute, ctx, req);
+    }
+
+    private String[] getLabelValues(RequestLog log) {
+        return ImmutableSortedMap.copyOf(labelingFunction.apply(log))
+                                 .values()
+                                 .stream()
+                                 .toArray(String[]::new);
+    }
+
+    private static boolean isSuccess(RequestLog log) {
+        final Object responseContent = log.responseContent();
+        if (responseContent instanceof RpcResponse) {
+            return !((RpcResponse) responseContent).isCompletedExceptionally();
+        }
+
+        if (log.responseCause() != null) {
+            return false;
+        }
+
+        if (HttpSessionProtocols.isHttp(log.sessionProtocol())) {
+            if (log.statusCode() >= 400) {
+                return false;
+            }
+        } else {
+            if (log.statusCode() != 0) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    @FunctionalInterface
+    private interface ThrowingBiFunction<C, I, O> {
+        O apply(C c, I i) throws Exception;
+    }
+
+    private static final class MetricAdapter<T extends MetricLabel> {
+        private final Summary timer;
+        private final Counter success;
+        private final Counter failure;
+        private final Gauge activeRequests;
+        private final Summary requestBytes;
+        private final Summary responseBytes;
+
+        private MetricAdapter(PrometheusRegistryWrapper collectorRegistry,
+                              T[] metricLabels,
+                              String namePrefix,
+                              String help) {
+            Objects.requireNonNull(collectorRegistry, "collectorRegistry");
+            Objects.requireNonNull(metricLabels, "metricLabels");
+            Objects.requireNonNull(namePrefix, "namePrefix");
+            Objects.requireNonNull(help, "help");
+
+            final String[] metricLabelNames = Stream.of(metricLabels)
+                                                    .sorted()
+                                                    .map(MetricLabel::name)
+                                                    .toArray(String[]::new);
+            timer = collectorRegistry.create(namePrefix + "_request_duration_nanoseconds",
```

Comment:
I'm actually curious if it's just fine to use nanoseconds, given that we may lose some precision. Losing precision may not really be a problem though because it's time.